### PR TITLE
Changes grids to be symmetric for cleaner FFTs

### PR DESCRIPTION
**Pull request by mturaev**
Grids are now symmetric which simplifies ft_shift  and ft^(-1)_shift. Modified convolutions, cross correlations and Hilbert to work with this new grid discritisation. Various notational changes which are hopefully more intuitive .

### CODE REVIEW

File: src/grids/linear_grid.jl [hunk 1]
```diff
@@ -1,217 +1,218 @@
...
-  unsafe_execute!(ft, _zeropad!(padded, u), out)
-  @. out *= ft_shift
-  
-  return out
+function ft!(grid::LinearGrid, out::AbstractVector, u::AbstractVector)
+    return fftshift(bfft(u)) .* grid.op.ft_arr[1:length(u)] * grid.d
 end
 
-fft(grid::LinearGrid, u::AbstractVector) = fft!(grid, zero(grid.op.ft_work1), u)
+ft(grid::LinearGrid, u::AbstractVector) = ft!(grid, zero(u), copy(u))
```
Comment (marvinlenk) [inline src/grids/linear_grid.jl:115]: Why the copy?

File: src/grids/linear_grid.jl [hunk 2]
```diff
@@ -1,217 +1,218 @@
...
 
 """
 Apply inverse FFT of grid to input
 """
-ifft!(grid::LinearGrid, out::AbstractVector, u::AbstractVector
-  ) = out .= @views _ifft!(grid, grid.op.ft⁻¹_work, u)[1:grid.len]
 
-ifft(grid::LinearGrid, u::AbstractVector
-  ) = ifft!(grid, zeros(eltype(grid.op.ft⁻¹_work), grid.len), u)
+ift(grid::LinearGrid, u::AbstractVector) = ift!(grid, zero(u), copy(u))
```
Comment (marvinlenk) [inline src/grids/linear_grid.jl:128]: Why the copy? See above.

File: src/grids/linear_grid.jl [hunk 3]
```diff
@@ -1,217 +1,218 @@
...
 """
 crosscorrelation!(grid::LinearGrid, out::AbstractVector,
-  u::AbstractVector, v::AbstractVector
-  ) = out .= @views _crosscorrelation!(grid, grid.op.ft⁻¹_work, u, v)[1:grid.len]
+    u::AbstractVector, v::AbstractVector
+) = out .= @views _crosscorrelation!(grid, grid.op.bft_work, u, v)[grid.op.data_range]
 
 crosscorrelation(grid::LinearGrid, u::AbstractVector, v::AbstractVector
-  ) = crosscorrelation!(grid, zero(u), u, v)
+) = crosscorrelation!(grid, zeros(eltype(grid.op.padded), length(u)), u, v)
```
Comment (marvinlenk) [inline src/grids/linear_grid.jl:187]: We should make both u and v have the same types AbstractVector{T}. But The output should have the same data structure and size as the input, no? This function is on non-padded arrays.

File: src/grids/linear_grid.jl [hunk 4]
```diff
@@ -1,217 +1,218 @@
...
+
+    assert_applicable(bft_p, ft_work2, out)
+    unsafe_execute!(ft_p, _zeropad!(padded, u), ft_work1)
+    unsafe_execute!(ft_p, _zeropad!(padded, v), ft_work2)
+
+    @. ft_work2 *= conj(ft_work1) * ft_arr
+    unsafe_execute!(bft_p, ft_work2, out)
+
+
+    out .*= grid.d / length(padded)
```
Comment (marvinlenk) [inline src/grids/linear_grid.jl:174]: Performance wise it should be much better to multiply this into ft_work2 before the FT, I don't know if this is bad for the data though. Should not be, but who knows.

File: src/grids/linear_grid.jl [hunk 5]
```diff
@@ -1,217 +1,218 @@
...
+    (; padded, ft_work1, ft_work2, ft_p, bft_p, ft_arr) = grid.op
+
+    assert_applicable(bft_p, ft_work2, out)
+    unsafe_execute!(ft_p, _zeropad!(padded, u), ft_work1)
+    unsafe_execute!(ft_p, _zeropad!(padded, v), ft_work2)
+    @. ft_work2 *= ft_work1 * ft_arr
+
+    unsafe_execute!(bft_p, ft_work2, out)
+
+    out .*= grid.d / length(padded)
```
Comment (marvinlenk) [inline src/grids/linear_grid.jl:144]: Performance wise it should be much better to multiply this into ft_work2 before the FT, I don't know if this is bad for the data though. Should not be, but who knows.

File: src/grids/linear_grid.jl [hunk 6]
```diff
@@ -1,217 +1,218 @@
...
-  
-  return out
+    (; padded, ft_work1, ft_p, bft_p, fft_x⁻¹) = grid.op
+
+    assert_applicable(bft_p, ft_work1, out)
+    unsafe_execute!(ft_p, _zeropad!(padded, u), ft_work1)
+    @. ft_work1 *= fft_x⁻¹
+    unsafe_execute!(bft_p, ft_work1, out)
+
+    out .*= 1 / length(padded)
```
Comment (marvinlenk) [inline src/grids/linear_grid.jl:200]: See convolution.

File: src/grids/linear_grid.jl (hunk 1 above)
Comment (mturaev) [reply src/grids/linear_grid.jl:115]: To be honest, it is not necessary. 
We have to multiply u by ft_arr before we fourier transform, so we either just multiply (not in place) u by ft_arr before fft or like I did here create a copy which is then multiplied in place. 
Maybe there is a smarter way?

File: src/grids/linear_grid.jl (hunk 3 above)
Comment (mturaev) [reply src/grids/linear_grid.jl:187]: yes, this needs to be done in more detail. If one inputs a real array u, due to complex_op = true by default, the output must be complex. This was just a quick fix..

File: src/grids/linear_grid.jl (hunk 1 above)
Comment (marvinlenk) [reply src/grids/linear_grid.jl:115]: In this specific case it actually just calls bfft on u, which should be non-altering. But yes, when we use the plan in the future we need to think about this actually.

File: src/grids/linear_grid.jl (hunk 3 above)
Comment (marvinlenk) [reply src/grids/linear_grid.jl:187]: Right, we should check actually what happens when we give a real function to a complex_op=true grid in general and make this stable.

**Review by marvinlenk — APPROVED** [review]
Looks nice. The remarks can be addressed in a future version.

**Pull request merged**